fix memory leaks and thread issues in BCFitter and descendants

fredRos opened the pull request:
I managed to make the code much safer and removed a total of 500 lines.
The first commit is just convenience to avoid memory leaks. Check the message of the second commit for an overview of changes.  `make check` passed with multiple threads on root 5.34.25, 6.04.10 and 6.05.02

oschulz commented:
Tests run fine on OS-X but on Linux, tests fail reproducibly at "graphFitterAdvancedExample.C" when using ROOT-5.34.32. With ROOT-6, all seems fine.

fredRos commented:
I fixed `graphFitterAdvancedExample.C` for root 5, updated the example/README to resolve #77 and converted it to markdown, and added some more minor improvements and code polishing. We are getting there...

oschulz commented:
Now distcheck fail completely ;-)

    [...]
    make[2]: Entering directory `[...]/bat/examples'
    make[2]: *** No rule to make target `README', needed by `distdir'.  Stop.

Something missing in dist?

oschulz commented on BAT/BCAux.h:
```diff
@@ -119,6 +119,47 @@ void DrawKnowledgeUpdate(BCHistogramBase& prior, BCHistogramBase& posterior, boo
+ * transferred in copies, only during a swap.
+ */
+template <typename T>
+class BCTrash
```
BCTrash might not be an optimal name for this class:

    BCTrash.Put(...)

sounds like "delete this", instead of "keep this alive". :-)

fredRos replied:
> BCTrash might not be an optimal name for this class:
> 
> BCTrash.Put(...)
> 
> sounds like "delete this", instead of "keep this alive". :-)

In fact Dan and I tossed a lot of different names up, this one remained. Think about this way: if you want a file removed without getting it back you do `rm file`. If you put it into the trash, you know it's still there until somebody empties the trash later. This corresponds pretty well to our use case at hand.

If, however, you can suggest a better name, I'm happy to change it again.

oschulz replied:
I don't have a really good name for it, either. Well, it's an internal class, so we can change it later, if we want to.
